pkg/oc/admin/diagnostics/diagnostics/cluster/app_create/main.go

```diff
@@ -0,0 +1,357 @@
+		{"route-port", "Router port to use for route connection test", &d.routePort, 80},
+		{"deploy-timeout", "Seconds to wait for the app to be ready", &d.deployTimeout, AppCreateTimeoutDefault},
+		{"admission-timeout", "Seconds to wait for the route to be admitted by a router", &d.routeAdmissionTimeout, AppCreateRouteAdmissionTimeoutDefault},
+		{"skip-service-connect", "Do not test connecting to the service", &d.skipServiceConnect, false},
+		{"skip-route-test", "Do not test route at all", &d.skipRouteTest, false},
+		{"skip-route-connect", "Do not test connecting to the route", &d.skipRouteConnect, false},
+		{"http-timeout", "Milliseconds to wait for an HTTP request to the app", &d.httpTimeout, AppCreateHttpTimeoutDefault},
+		{"http-retries", "Number of times to retry an HTTP request to the app", &d.httpRetries, AppCreateHttpRetriesDefault},
+		{"node-selector", "Node selector for where the test app should land", &d.nodeSelector, ""},
+		{"keep-app", "Do not delete the test app when complete", &d.keepApp, false},
+		{"result-dir", "Directory in which to write result details if desired", &d.writeResultDir, ""},
+	}
+}
+
+func getDefaultAppImage() string {
+	template := variable.NewDefaultImageTemplate()
+	return template.ExpandOrDie("deployer")
+}
+
+func (d *AppCreate) Complete(logger *log.Logger) error {
+	// project management
+	d.keepProject = d.keepProject || d.keepApp // don't delete project if keeping app
+	if d.project == "" && d.projectBase == "" {
+		return fmt.Errorf("%s project name cannot be empty", AppCreateName)
+	}
+	if d.project == "" {
+		// generate a project if not specified
+		d.project = names.SimpleNameGenerator.GenerateName(d.projectBase)
+	} else {
+		// when an existing project is specified, deleting it is likely to surprise the user, so don't
+		d.keepProject = true
+	}
+	if errs := kvalidation.IsDNS1123Label(d.project); len(errs) > 0 {
+		return fmt.Errorf("invalid project name '%s' for AppCreate: %v", d.project, errs)
+	}
+	// TODO: also test that route is valid under DNS952
+
+	// app management
+	if d.appName == "" {
+		return fmt.Errorf("%s app name cannot be empty", AppCreateName)
+	}
+	if errs := kvalidation.IsDNS1123Label(d.appName); len(errs) > 0 {
+		return fmt.Errorf("invalid app name '%s' for AppCreate: %v", d.appName, errs)
+	}
+	if err := kvalidation.IsValidPortNum(d.appPort); err != nil {
+		return fmt.Errorf("invalid app port %d for AppCreate: %v", d.appPort, err)
+	}
+	d.label = map[string]string{"app": d.appName}
+	d.labelSelector = fmt.Sprintf("app=%s", d.appName)
+
+	d.skipRouteConnect = d.skipRouteConnect || d.skipRouteTest // don't try to connect to route if skipping route test
+
+	return nil
+}
+
+func (d *AppCreate) CanRun() (bool, error) {
+	if d.SARClient == nil || d.AppsClient == nil || d.KubeClient == nil || d.ProjectClient == nil || d.RoleBindingClient == nil {
+		return false, fmt.Errorf("missing at least one client")
```

Comment:
Any chance to clarify a bit more the failure here? As a user I would not really know how to interpret this / how to remedy this situation

Reply:
Pretty much all the diagnostics have this pattern in place, established early on by deads. It's a developer failsafe to ensure there isn't a code path (e.g. unit test) that leaves out a requirement; if the user ever sees it, something has gone seriously wrong and there's no message that can help them except maybe "whip the developer".

Reply:
This will rarely be a user problem. In majority of cases this error will be only useful for developer, b/c he screwed the configuration somehow.